### ISSUE TITLE
Database schema improvments

### ISSUE THREAD

**Issue by blitzmann** [pyfa-org/Pyfa#961]
Starting a list to determine what we can do about database schema (specifically user database) to improve performance / relational integrity.

* Add `sqlite_autoincrement=True` to tables which have a primary key (maybe not all tables). This will prevent SQLite from reusing previous IDs, which has been a thorn in our side with database corruption for years. See http://docs.sqlalchemy.org/en/latest/dialects/sqlite.html#sqlite-auto-incrementing-behavior.
* Add CreateDate and ModifyDate columns to most tables, and have a SQLAlchemy hook that updates these on every record INSERT / UPDATE. This will help trouble shoot when things might happen. Need to look into how we could do this and what kind of size increase this instills on the database.
* Discuss column names. I've always hated how everything is ID instead of CharacterID or FitID. I also am partial to microsofts DB naming conventions (TitleCase). This would be a much bigger change if we were top change the column names, as we would have to ensure that everything that uses them uses the properly. As a stopgap, we could have aliases for them that are used in the code, but that could become messy as well. 

None of this is set in stone or planned, just throwing it out there for tracking.

**Comment by Ebag333** [comment]
>Discuss column names. I've always hated how everything is ID instead of CharacterID or FitID. 

It depends on which use case you're talking about.

Take `Boosters`.  You have:
`ID, itemID, fitID, active`

This is a good table design.  You _don't_ want:
`boosterID, itemID, fitID, active`

because other tables will use `boosterID` to reference this table.  So by this convention, `ID` is always self referential (and _always_ increments, never null, etc) while `boosterID` is referencing a foreign table, and shouldn't increment, might or might not be null, etc.

It becomes redundant when you're referencing it if you do it the second way because you will end up with something akin to:
`WHERE foreigntable.boosterID = booster.boosterID`

If you do it the way `boosters` is currently laid out then you have:
`WHERE foreigntable.boosterID = booster.ID`

Which is much more readable IMHO.  It's also a more common convention that I see very often.


Now, speaking of ID's you have tables like `characterSkills` which drives me nuts.  There's no unique ID in this table, which is problematic on a number of levels.  So this table should be changed from:
`characterID, itemID, _Skill__Level`
to:
`ID, characterID, itemID, skillLevel`

Where `ID` auto increments.  Yes our code doesn't care about the self referential ID on characterSkills, but it's a standard practice and we can do a number of things with it (such as checking for duplicates more easily).  When querying counts, currently you have to do this:
`select count(*) form characterSkills WHERE <condition>`
with an ID column you can do:
`select count(ID) from characterSkills WHERE <condition>`

The second is a lower load on the database and a faster query.  Granted our tables aren't tens of thousands so the difference is minuscule, but it does add up.


Speaking of foreign keys, you should add that to your list.  And if we're writing foreign keys, then they're easier to write and troubleshoot when one side is the same.

For example:
````
CREATE TABLE fits(
<snip>
FOREIGN KEY(damagePatternID) REFERENCES damagePattern(ID)
)
```
is easier to read than:
````
CREATE TABLE fits(
<snip>
FOREIGN KEY(damagePatternID) REFERENCES damagePattern(damagePatternID)
)
```

So TL:DR;
A `ID` column on each table is a good thing because it's easier to make very standardized, while virtually every other column will have special cases surrounding it.  This way we can quickly identify non-standard `ID` columns, where as if there are two `boosterID` columns it's harder to tell which is which.



The other two  things you mention I'm 100% on board with.  Especially the created/modified dates, those are super useful for troubleshooting.

**Comment by Ebag333** [comment]
>we should have a table for table versions

Something else to add.

**Comment by blitzmann** [comment]
> because other tables will use boosterID to reference this table. So by this convention, ID is always self referential (and always increments, never null, etc) while boosterID is referencing a foreign table, and shouldn't increment, might or might not be null, etc.

I feel like this is preferential. In my job, we always have a column that is basically `<tableName>ID` and I find it much more intuitive.

> Which is much more readable IMHO. 

I'm completely the opposite. Then again, this isn't something I am very set on. I will be reading some literature soon about common conventions when it comes to these things and just go from there.

**Comment by Ebag333** [comment]
>I feel like this is preferential. 

It really is. There's no technical reason for choosing one over the other. 

I think that the main advantage is we can have standard structural columns (that all look/act the same regardless of the table, such as ID, date modified, date created, etc) and then the data columns that hold data or point to other tables. 

Some databases get crazy with that, with the column headers being named something like `fkFit` for a reference to another table. I don't think we need to go that far. :)

**Comment by blitzmann** [comment]
Took a look at this tonight. I'm not going to worry about the `ID` vs `TableID` stuff, or changing to TitleCase. Way too much work for no benefit other than that's what I'm used to for other projects.

However, the create/modify stuff should be super simple:
```
import sqlalchemy.sql.functions as func

...

Column("created", DateTime, nullable=True, default=func.now()),
Column("modified", DateTime, nullable=True, onupdate=func.now()),
...
```

We will have to use a migration script to alter the existing tables to include this ofc.

The `sqlite_autoincrement` might be a bit more challenging - SQLite doesn't support in-place table alterations, so doing an `ALTER TABLE table ALTER/MODIFY COLUMN column` won't work. The official way to do this would be to move all data to temp tables, drop the existing tables, recreate with correct schema, reload the data, and then drop the temp tables. There's another way to do it using a PRAMA command and editing the sqlite master table definition. It's generally recommended not to do this, but this is going to be a tested automated thing on a schema we know about, so I'm thinking we can use it safely. I'm going to look more into this though to make sure.

**Comment by Ebag333** [comment]
I actually started a branch on this. Auto ID is easy. Date created is easy. Haven't tested onupdate but looks easy enough. 

Basically so far I had a migration script started, but hadn't touched the table definitions.

**Comment by blitzmann** [comment]
Erm, already made a pull request. Will be continuing to test it out the next few days, please give it a whirl as well :D

> Auto ID is easy. 

I haven't start on this bit though, if you're interested in fixing up that branch to only include this bit and make a PR, we can collaborate on getting this in there. How did you go about doing it?

**Comment by Ebag333** [comment]
Rename existing table, create new table, migrate data, delete old table. Easy peasy. :) 

Aw crap. I think that I purged it when I did my last branch cleanup. 

I can recreate it easy enough, the SQLite DB tool creates the migration script for you, just need to convert it to SQLAlchemy/Python.

I would really recommend we standardize the tables to an
`ID | <data> | date_modified | date_created`
format. 

Where <data> is all the references to other tables and misc information, and ID is the self referential auto incremental column. Half the tables already have an ID column so it's pretty easy.
